icu4c/source/common/uresbund.cpp

```diff
@@ -2950,7 +2950,15 @@ ures_getKeywordValues(const char *path, const char *keyword, UErrorCode *status)
     valuesBuf[0]=0;
     valuesBuf[1]=0;
     
-    while((locale = uenum_next(locs, &locLen, status)) != 0) {
+    // res_index.txt does not include "root" but we should also include the
+    // keyword values which only coded in "root" to the result, such as "emoji"
+    // and "eor" for "coll".
+    bool addRoot = true;
```

Comment:
I think that a simpler way of doing this would be to instead rewrite the loop like this:

```C++
    static constexpr char ROOT[] = "root";
    do {
        locale = uenum_next(locs, &locLen, status);
        if (locale == NULL) locale = ROOT;

        // ...
        
    } while (locale != ROOT);
```

That avoids calling `uenum_next()` one superfluous time after it already has reached the end and I think it will also be more clear to the human reader what the loop does.

Reply:
I found a better fix

Reply:
sorry, after playing with ICU4J I realize my attempted fix was wrong. Instead, I should just use res_open intead of res_openDirect and the loop will look into the fallback resources inside "root" and do not need to look into one more resource after the loop. PTAL. Noitce the only calling ures_getKeywordValues is inside 
ucol_getKeywordValues

which is where we intend to fix and the only internal place which call that function is Collator::getKeywordValues. so there are no risk to change this code for other  API.

Reply:
I suspect that it used to call ures_openDirect() as an optimization: ures_open() also opens all parent bundles down to root, but since the code here visits all locales, including the parent bundles, it will see their contents even with ures_openDirect(). Except if the "all locales" skips root, and if parent bundles for some reason are not listed as "available". So this change will make the operation slower, but also more reliably correct.